### DISCUSSION TITLE
Parallel request overlap service session

### DISCUSSION THREAD

**Discussion by Kokoroou** [fastapi/fastapi#12539] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from dataclasses import dataclass
import uvicorn
from fastapi import FastAPI, Request


@dataclass
class ServiceSession:
    request: Request

class Service:
    def create_problem(self):
        x1 = id(self.session)
        print(x1)
        x2 = id(self.session)
        print(x1 == x2)  # Mostly True, but False may appear in some cases

    def __call__(self, request: Request):
        self.session = ServiceSession(request=request)
        self.create_problem()
        return {"key": self.session.request.state.key}

service = Service()
app = FastAPI()

@app.get("/test")
def func_test(request: Request):
    request.state.key = request.headers.get("key")
    return service(request=request)

if __name__ == "__main__":
    uvicorn.run(app)
```

### Description

In the above example, I expect service session remain separated between parallel requests, so their location in memory is not change. But just some code between (`print(x1)`) can make the session of a request change to another one.

I test with this code to compare the expected value with the value received from API

```python
import concurrent.futures

import requests

URL = "http://localhost:8000/test"
MAX_ATTEMPTS = 50

# Function to handle each request
def attempt_request(value):
    response = requests.get(URL, headers={"key" : value})
    result = response.json()
    return value, result["key"]

# Running the requests in parallel
print("EXPECT --> RESULT")
with concurrent.futures.ThreadPoolExecutor() as executor:
    futures = [
        executor.submit(attempt_request, f"test_{i}")
        for i in range(MAX_ATTEMPTS)
    ]
    for future in concurrent.futures.as_completed(futures):
        original_variable, new_variable = future.result()
        if original_variable != new_variable:
            print(original_variable, "-->", new_variable, "MISMATCH")
        else:
            print(original_variable, "-->", new_variable)
```

And the result is like below
```
EXPECT --> RESULT
test_2 --> test_2
test_0 --> test_0
test_6 --> test_6
test_4 --> test_3 MISMATCH
test_7 --> test_7
test_5 --> test_5
test_8 --> test_8
test_1 --> test_1
test_3 --> test_5 MISMATCH
test_9 --> test_9
test_10 --> test_10
...
```

I can temporarily fix by initializing a Service for each request, but not sure what the problem is. Do you know what is the problem here and how to fix it permanently?

### Operating System

Linux

### FastAPI Version

0.111.0

### Pydantic Version

2.9.1

### Python Version

3.10.15

**Comment by luzzodev** [comment, marked as answer]
When you use a sync def
``` python
@app.get("/test")
def func_test(request: Request):
```
Fastapi will call it spawning a new thread by calling await `run_in_threadpool`

Your `class Service` is not thread-safe so one request could override the other.
You can try to use a `TLock` on `self.session=`

**Reply by YuriiMotov** [reply]
Actually, It doesn't depend on using `async def` or `def`. It will happen in any case.

`service` is global variable and this behavior is not surprising at all.

I'm not sure I understand the purpose of this construction, but creating new `Service` object on every request looks like a good approach.
You can create a dependency for this reason:

```
from dataclasses import dataclass
import time
from typing import Annotated
import uvicorn
from fastapi import Depends, FastAPI, Request


@dataclass
class ServiceSession:
    request: Request

class Service:
    def __init__(self, request: Request):
        self.session = ServiceSession(request=request)

    def dont_create_problem(self):
        x1 = id(self.session)
        print(x1)
        # time.sleep(1)  # You can even add sleep
        x2 = id(self.session)
        print(x1 == x2)  # Always True



app = FastAPI()

async def get_service(request: Request):
    return Service(request=request)

@app.get("/test")
def func_test(request: Request, service: Annotated[Service, Depends(get_service)]):
    request.state.key = request.headers.get("key")
    service.dont_create_problem()
    return {"key": service.session.request.state.key}


if __name__ == "__main__":
    uvicorn.run(app)

```

**Reply by luzzodev** [reply]
Never said it depends on sync or async.
The example provided is clearly a scenario where a shared resource is accessed or modified in an unsafe manner, and a race condition clearly occurs in both sync and async cases. To avoid abnormal behavior due to this race condition, you can safely use a lock.
Anyways in general accessing to shared resources should be avoided and other options should be preferred if there any.
In his case if create a new service for each request does not cost too much than this option should be preferred instead accessing to shared resource with locking

**Reply by YuriiMotov** [reply]
Sorry for misunderstanding

**Reply by Kokoroou** [reply]
In my case, Service is used to serving an AI model, so recreate it everytime is not an ideal solution. With small model, it can work without caring, but for large model, it can consume a lot of memories and take times to initialize. Set AI model as a global attribute can lead to the race condition too. So, I will try using `TLock` to see if it can fix my problem. Anyways, thanks a lot for replying!
